### PR TITLE
Add docstrings and refactor API

### PR DESCRIPTION
**Pull request by mjdietzx**
- Added JSDoc docstring validation to eslint rules.
- Added eslint npm modules to project's dev dependencies.
- Added docstrings for all public functions in the API (excluding `adapter.js` in this PR).
- Fixed eslint errors/warnings for all files in API
- Minor refactors and cleanup

Note: excluding `adapter.js` in this PR b/c there are so many changes there - PR would be huge. Therefore breaking PR into two.

### CODE REVIEW

File: api/descriptor.js [hunk 1]
```diff
@@ -39,16 +39,35 @@
...
+            throw new Error('uuid must be provided.');
+        }
 
         // if (!value) throw new Error('value must be provided.');
 
-        this._instanceId = characteristicInstanceId + '.' + (i++).toString();
+        // increment global so `descriptorInstanceId` is unique for each created service.
+        i += 1;
+
+        this._instanceId = `${characteristicInstanceId}.${(i).toString()}`;
```
Comment (mrodem) [inline api/descriptor.js]: Use of `.toString()` is not needed here anymore.

File: api/characteristic.js [hunk 2]
```diff
@@ -39,15 +39,37 @@
...
+        }
+        if (!properties) {
+            throw new Error('properties must be provided.');
+        }
+
+        // increment global so `characteristicInstanceId` is unique for each created service.
+        i += 1;
 
-        this._instanceId = serviceInstanceId + '.' + (i++).toString();
+        this._instanceId = `${serviceInstanceId}.${(i).toString()}`;
```
Comment (mrodem) [inline api/characteristic.js]: Use of `.toString()` is not needed here anymore.

File: api/service.js [hunk 3]
```diff
@@ -39,17 +39,31 @@
...
+     * @param {string} deviceInstanceId The unique ID of the Bluetooth device to add `Service` to.
+     * @param {string} uuid A 128-bit or 16-bit unique identifier for this service.
+     * @param {string} type The server service type. 'primary' or `secondary`.
+     */
     constructor(deviceInstanceId, uuid, type) {
-        this._instanceId = deviceInstanceId + '.' + (i++).toString();
+        // increment global so `serviceInstanceId` is unique for each created service.
+        i += 1;
+
+        this._instanceId = `${deviceInstanceId}.${(i).toString()}`;
```
Comment (mrodem) [inline api/service.js]: Use of `.toString()` is not needed here anymore.

File: api/serviceFactory.js [hunk 4]
```diff
@@ -43,35 +43,92 @@ const Service = require('./service');
...
-    createService(uuid, serviceType) {
-        if (!serviceType) serviceType = 'primary';
+    /**
+     * Factory method to create a service in the Bluetooth `Device's` 'local.server' GATT attribute table.
+     *
+     * @param {string} uuid A 128-bit or 16-bit unique identifier for this service.
+     * @param {string} serviceType The server service type. 'primary' (default) or `secondary`.
+     * @returns {Service} A newly created `Service` instance.
+     */
+    createService(uuid, serviceType = 'primary') {
```
Comment (mrodem) [inline api/serviceFactory.js]: See earlier comment about default function paramenters and Node.js v4.

File: api/serviceFactory.js [hunk 5]
```diff
@@ -43,35 +43,92 @@ const Service = require('./service');
...
+     * - {number} maxLength: Maximum attribute value length in bytes, see ref BLE_GATTS_ATTR_LENS_MAX for maximum values.
+     * - {boolean} readAuth: Read authorization and value will be requested from the application on every read operation.
+     * - {boolean} writeAuth: Write authorization will be requested from the application on every Write Request operation (but not Write Command).
+     * Note: vloc = this._bleDriver.BLE_GATTS_VLOC_STACK; // Attribute Value is located in stack memory, no user memory is required.
+     * @param {Object} options This GATT characteristic's attribute's metadata.
+     * @returns {Characteristic} A newly created `Characteristic` instance.
+     */
     createCharacteristic(service, uuid, value, properties, options) {
-        if (!service) throw new Error('Service to add characteritics to must be provided.');
+        // TODO: we are mutating the service parameter here. May want to re-work this method.
```
Comment (mrodem) [inline api/serviceFactory.js]: Good observation :+1:

**Comment by bihanssen** [comment]
:+1:

**Pull request merged**